### DISCUSSION TITLE
docker run options for network specification

### DISCUSSION THREAD

**Discussion by sahanajoshi1996** [gradescope/autograder_samples#86] [Q&A] [answered]
Hi,
I want to setup traffic control for a programming assignment using Linux's [tc](https://man7.org/linux/man-pages/man8/tc-netem.8.html) tool. I was able to install the package but I'm unable to run the command to setup rate limiting `tc qdisc add dev lo root netem delay 1ms rate 1000kbit limit 40`. 

I get an error `RTNETLINK answers: Operation not permitted`, which upon looking up in docker platforms, I found out that it's because a container does not have network admin [privileges](https://forums.docker.com/t/cannot-start-wireguard-server-rtnetlink-answers-operation-not-permitted/119208/5). 

One tested way to fix this error is by adding `--cap-add=NET_ADMIN` option to docker run command when starting the container. But I don't know how I could do that with Gradescope, because I can only control the Dockerfile and not the container creation itself. Any help is greatly appreciated! :)

Note: I ensured that tc is correctly installed by running a show command and that my user was root(as it should be inside a container), so I know there are no permission issues, the problem is the container does not have permission to modify the interface.

**Comment by ibrahima** [comment]
Hi, unfortunately it is not possible to change the Linux capabilities of the Docker container or run in privileged mode on Gradescope at this time. One approach that we have suggested in the past is emulating your own system inside our Docker containers using QEMU but this will probably make things very slow - I don't know if it would become slow to the point of being unusable though. If you do try that approach please let us know how that goes! Sorry I don't have a better answer for you right now.

**Reply by sahanajoshi1996** [reply]
Hi, 

Thank you for your prompt response!
I'm considering the QEMU option, but will that not require setting device options such as `device==/dev/kvm` ? And because we have no control on how the container is started, i.e, no control over `run` command(at least not that I know of :) ), how can we use QEMU?

Thanks again! :)

**Reply by sahanajoshi1996** [reply]
I guess my question is if there is a way I can use any of these Linux capability [options](https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities) to start a container on Gradescope.

**Reply by ibrahima** [reply, marked as answer]
No, it's not possible to use those options. I think you can run QEMU without additional privileges but I am not 100% sure.
